mixnet/config.py

```diff
@@ -2,110 +2,77 @@
-
-    def __init__(
-        self,
-        config: MixnetTopologyConfig,
-    ) -> None:
-        """
-        Build a new topology deterministically using an entropy and a given set of candidates.
-        """
-        shuffled = FisherYates.shuffle(config.mixnode_candidates, config.entropy)
-        sampled = shuffled[: config.size.num_total_mixnodes()]
+class MixMembership:
+    """
+    A list of public information of nodes in the network.
+    We assume that this list is eventually known to all nodes in the network (e.g. via p2p advertising).
+    """
 
-        layers = []
-        for layer_id in range(config.size.num_layers):
-            start = layer_id * config.size.num_mixnodes_per_layer
-            layer = sampled[start : start + config.size.num_mixnodes_per_layer]
-            layers.append(layer)
-        self.layers = layers
+    nodes: List[NodeInfo]
 
-    def generate_route(self, mix_destination: MixNodeInfo) -> list[MixNodeInfo]:
+    def generate_route(self, num_hops: int, last_mix: NodeInfo) -> list[NodeInfo]:
         """
         Generate a mix route for a Sphinx packet.
         The pre-selected mix_destination is used as a last mix node in the route,
         so that associated packets can be merged together into a original message.
         """
-        route = [random.choice(layer) for layer in self.layers[:-1]]
-        route.append(mix_destination)
-        return route
+        return [*(self.choose() for _ in range(num_hops - 1)), last_mix]
 
-    def choose_mix_destination(self) -> MixNodeInfo:
+    def choose(self) -> NodeInfo:
         """
-        Choose a mix node from the last mix layer as a mix destination
-        that will reconstruct a message from Sphinx packets.
+        Choose a mix node as a mix destination that will reconstruct a message from Sphinx packets.
         """
-        return random.choice(self.layers[-1])
+        return random.choice(self.nodes)
 
 
 @dataclass
-class MixnetTopologySize:
-    num_layers: int
-    num_mixnodes_per_layer: int
-
-    def num_total_mixnodes(self) -> int:
-        return self.num_layers * self.num_mixnodes_per_layer
-
-
-# 32-byte that represents an IP address and a port of a mix node.
-NodeAddress: TypeAlias = bytes
-
-
-@dataclass
-class MixNodeInfo:
-    identity_private_key: BlsPrivateKey
-    encryption_private_key: X25519PrivateKey
-    addr: NodeAddress
-
-    def identity_public_key(self) -> BlsPublicKey:
-        return self.identity_private_key.get_g1()
+class NodeInfo:
+    """
+    Public information of a node to be shared to all nodes in the network
+    """
 
-    def encryption_public_key(self) -> X25519PublicKey:
-        return self.encryption_private_key.public_key()
+    public_key: X25519PublicKey
 
-    def sphinx_node(self) -> Node:
-        return Node(self.encryption_private_key, self.addr)
+    def sphinx_node(self) -> SphinxNode:
+        # TODO: Use a pre-signed incentive tx, instead of NodeAddress
```

Comment:
This is from the Alvaro's design proposal: https://www.notion.so/Maximally-decentralized-and-homogeneous-private-consensus-6df631b1c7ad430e8846f64fb874ccda?pvs=4#53ef5eb6d7ff447499e20562551f1507
> Every layer of the message has an additional field: a pre-signed transaction that can be redeemed by the receiver to incentivize relaying (ie mixing).

I added this comment before we meet in the offsite. If this is not the case in our design, of if it is not determined yet (I guess), I can remove this comment.

Reply:
Incentivization is not part of the design so it might be confusing for the reader. It would be better to remove it and introduce it again when we design that part also.

Reply:
I agree. I just removed the comment: e964721